src/common/database.cpp

```diff
@@ -0,0 +1,117 @@
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  This program is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program.  If not, see http://www.gnu.org/licenses/
+
+===========================================================================
+*/
+
+#include "database.h"
+
+#include "logging.h"
+#include "settings.h"
+#include "utils.h"
+
+namespace
+{
+    mutex_guarded<db::detail::State> state;
+} // namespace
+
+mutex_guarded<db::detail::State>& db::detail::getState()
+{
+    TracyZoneScoped;
+
+    // TODO: Manual pooling?
+    // TODO: Locking of individual connections by passing back a handle with a `*this` reference to free the connection on handle destruction?
+
+    // clang-format off
+    if (state.read([&](const auto& state) { return state.connection != nullptr; }))
+    {
+        return state;
+    }
+
+    state.write([&](auto& state)
+    {
+        // NOTE: Driver is static, so it will only be initialized once.
+        sql::Driver* driver = sql::mariadb::get_driver_instance();
+
+        try
+        {
+            auto login  = settings::get<std::string>("network.SQL_LOGIN");
+            auto passwd = settings::get<std::string>("network.SQL_PASSWORD");
+            auto host   = settings::get<std::string>("network.SQL_HOST");
+            auto port   = settings::get<uint16>("network.SQL_PORT");
+            auto schema = settings::get<std::string>("network.SQL_DATABASE");
+            auto url    = fmt::format("tcp://{}:{}", host, port);
+
+            state.connection = std::unique_ptr<sql::Connection>(driver->connect(url.c_str(), login.c_str(), passwd.c_str()));
+            state.connection->setSchema(schema.c_str());
+        }
+        catch (const std::exception& e)
+        {
+            ShowError(e.what());
+            state.connection = nullptr; // Wipe the connection so that it can't be used if it's broken
+        }
+    });
+    // clang-format on
+
+    return state;
+}
+
+std::string db::detail::sanitise(std::string const& query)
+{
+    const auto replaceAll = [](std::string const& str, std::vector<std::string> const& from, std::string const& to)
+    {
+        std::string result = str;
+        for (auto const& substr : from)
+        {
+            size_t start_pos = 0;
+            while ((start_pos = result.find(substr, start_pos)) != std::string::npos)
+            {
+                result.replace(start_pos, substr.length(), to);
+                start_pos += to.length();
+            }
+        }
+        return result;
+    };
+
+    return replaceAll(query, { "--", "/*", "*/", ";" }, "");
```

Comment:
Isn't this supposed to replace "bad" characters with escaped ones? `'` -> `\'` and so on so that things like chat, search comments, bazaar comments, etc can still include them, but they also wreck injection?

Speaking of, `'` is missing from this list.

Reply:
see:
https://dev.mysql.com/doc/refman/8.0/en/string-literals.html

Reply:
also the autostranslate junk people can put in bazaar and seacom, thats not even human visible characters.

Reply:
> Isn't this supposed to replace "bad" characters with escaped ones? ' -> \' and so on so that things like chat, search comments, bazaar comments, etc can still include them, but they also wreck injection?
> 
> Speaking of, ' is missing from this list.

My blob encoding handles escaping characters and things, this sanitise routine is specifically to make super ultra mega sure we're removing any and all possible vectors for SQL injection, even if the c connector, cpp connector and other mechanisms all promise they stop it on our behalf

Reply:
autotranslate, bazaar, seacon etc. are all handled as blobs, which I now have acting equivalently to the c-connector, and it's now obvious what it's doing.

This routine is for the cases where plain text is pulled out of incoming packets and used as strings, rather than blobs for the db or the client

Reply:
are gm/chat audits blobs?

Reply:
Should be, easy test thanks!

Reply:
Will do this in later PRs, I already have evidence of a blob being read from the old system, rewritten and reread correctly